NumLib/ODESolver/PETScNonlinearSolver.h

```diff
@@ -0,0 +1,242 @@
+            DBUG("The petsc-x vector.")
+            VecView(petsc_x, PETSC_VIEWER_STDOUT_WORLD);
+            */
+
+            // Assemble in ogs context.
+            BaseLib::RunTime time_assembly;
+            time_assembly.start();
+            context->system->assemble(*context->x);
+
+            INFO("[time] Assembly took %g s.", time_assembly.elapsed());
+            context->system->getResidual(*context->x, *context->r);
+
+            context->system->getJacobian(*context->J);
+            context->system->applyKnownSolutionsNewton(*context->J, *context->r,
+                                                       *context->x);
+            /*
+            DBUG("AFTER ASSEMBLY");
+            DBUG("The ogs-x vector.")
+            VecView(context->x->getRawVector(), PETSC_VIEWER_STDOUT_WORLD);
+            DBUG("The ogs-r vector.")
+            VecView(context->r->getRawVector(), PETSC_VIEWER_STDOUT_WORLD);
+            */
+
+            // Now copy the results into petsc vectors.
+            VecCopy(context->r->getRawVector(), petsc_r);
+
+            /*
+            DBUG("The petsc-x vector.")
+            VecView(petsc_x, PETSC_VIEWER_STDOUT_WORLD);
+            DBUG("The petsc-r vector.")
+            VecView(petsc_r, PETSC_VIEWER_STDOUT_WORLD);
+            */
+            return 0;
+        };
+
+        auto jacobian = [](SNES /*snes*/,
+                           Vec /*petsc_x*/,
+                           Mat petsc_J,
+                           Mat /*petsc_B* Same as petsc_J*/,
+                           void* petsc_context) {
+            DBUG("PETScNonlinearSolver: jacobian callback called.")
+            // Assume the system is already assembled.
+            auto context = static_cast<::detail::PetscContext*>(petsc_context);
+            context->system->getJacobian(*context->J);
+
+            /*
+            DBUG("The ogs-J matrix.")
+            MatView(context->J->getRawMatrix(), PETSC_VIEWER_STDOUT_WORLD);
+            */
+            // Now copy the results into petsc vectors.
+            MatCopy(context->J->getRawMatrix(), petsc_J,
+                    DIFFERENT_NONZERO_PATTERN);
+            /*
+            DBUG("The petsc-J matrix.")
+            MatView(petsc_J, PETSC_VIEWER_STDOUT_WORLD);
+            */
+            return 0;
+        };
+
+        DBUG("PETScNonlinearSolver: set function");
+        SNESSetFunction(_snes_solver, petsc_r.getRawVector(), residual,
+                        &petsc_context);
+
+        DBUG("PETScNonlinearSolver: set jacobian");
+        // The jacobian and the preconditioner matrices point to the same
+        // location.
+        SNESSetJacobian(_snes_solver, petsc_J.getRawMatrix(),
+                        petsc_J.getRawMatrix(), jacobian, &petsc_context);
+
+        DBUG("PETScNonlinearSolver: set constraints");
+        // Constraints
+        Vec xl, xu;
+        VecDuplicate(x.getRawVector(), &xl);
```

Comment:
Frequently allocating/releasing memory for xl, xu, J, r, x may slow down the computation. It could be better to set these variables as the members of this class, and allocate memory for them only once.